### PR TITLE
Add Cannonsville and Pepacton diversion data

### PR DESCRIPTION
**Pull request by limnoliver**
This PR gathers and munges diversion (and other) data that was provided by the NY WSC (contact was Alicia Gearwar) in a variety of formats. This is how the data were stitched together:

- From 1980 to 2001, Alicia scanned reports for me, and I manually entered monthly diversion data from Pepacton and Cannonsville (that's all that was scanned). Monthly values are the average diversions per day in that month, so we can use these to populate daily values. The spreadsheet of manually extracted data is [here](https://doimspp-my.sharepoint.com/:x:/g/personal/soliver_usgs_gov/EeSBTOgDQuFOkFEREZS8qbABoJdgkoMIZwb7QDZ-pocmDA?e=xZd4lw). 
- From 2002 to 2008, there were online reports from which to manually pull monthly data (essentially the same reports that Alicia scanned for me, but available online). These reports were found [here](https://pubs.usgs.gov/wdr/#NY) and [here ](https://wdr.water.usgs.gov/allsearch.php)(note, these reservoirs are in E NY).
- From 2008 to present, the NY WSC had already manually pulled monthly values, and I munged the data from their spreadsheets (.xls files  in the main folder [here](https://doimspp-my.sharepoint.com/personal/soliver_usgs_gov/_layouts/15/onedrive.aspx?id=%2fpersonal%2fsoliver_usgs_gov%2fDocuments%2fTemperature+project+docs%2freservoir+data&FolderCTID=0x0120006D8ABDA28225104A99CB82266F822BB1)). Each file contains monthly data for one reservoir and one water year. Note these spreadsheets have more information than the manually pulled data, so starting in 2009 we have things like reservoir elevation, precipitation, runoff, etc. 
- From WY2009 to WY2020 (but missing some data in WY 2009), daily data are available in spreadsheets in the [WY folders, e.g., 2010](https://doimspp-my.sharepoint.com/personal/soliver_usgs_gov/_layouts/15/onedrive.aspx?FolderCTID=0x0120006D8ABDA28225104A99CB82266F822BB1&id=%2Fpersonal%2Fsoliver%5Fusgs%5Fgov%2FDocuments%2FTemperature%20project%20docs%2Freservoir%20data%2F2010WY%20data). In each water year folder, there is a .xls file per month, and reservoirs are stored in tabs. I only pulled Pepacton and Cannonsville data because I didn't want to deal with parsing the other reservoirs, which were sometimes in an inconsistent format.

Note, to build these data yourself, you need to download the entire [reservoir data](https://doimspp-my.sharepoint.com/:f:/g/personal/soliver_usgs_gov/EiawwIR59vdApLCkeyob4t0Bx86d2SoTic8ahCm6MH5wRA?e=kWC4jS) folder, and unzip into `2_observations/in`.

I "interpolated" daily values through the period of record (starting 1979-10-01) by 1) using daily values when they were available and 2) using the monthly values for every day of the corresponding month when daily values were missing.

```
diversion <- read_csv('2_observations/out/interpolated_daily_diversions.csv', col_types = "cDnnncnnnnnnnnnc")
ggplot(diversion, aes(x = date, y = diversion_cms)) + 
    geom_point(size = 0.1) + 
    facet_wrap(~reservoir, nrow = 2)
```
image

### CODE REVIEW

File: 2_observations.yml [hunk 1]
```diff
@@ -96,8 +97,37 @@ targets:
...
+
+  # interpolate to daily values
+  # use daily observed values when available, populate same diversion value
+  # over month when monthly values vailable
+  2_observations/out/interpolated_daily_diversions.csv.ind:
+    command: interpolate_to_daily(
+      out_ind = target_name,
+      daily_ind = '2_observations/out/observed_daily_diversions.csv.ind',
+      monthly_ind = '2_observations/out/observed_monthly_diversions.csv.ind',
+      mgd_to_cms = 0.0438, cfs_to_cms = 0.0283, mg_to_cm = 4404.88, ft_to_m = 0.3048)
```
Comment (aappling-usgs) [inline 2_observations.yml:132]: Wondering if these conversion factors really need to be passed as arguments visible to scipiper. They're not ever going to change, and they don't embody anything about our preferences for how the data are processed, right?

File: 2_observations.yml [hunk 2]
```diff
@@ -96,8 +97,37 @@ targets:
...
+  # you must download reservoir data from here: 'https://doimspp-my.sharepoint.com/:f:/g/personal/soliver_usgs_gov/EiawwIR59vdApLCkeyob4t0Bx86d2SoTic8ahCm6MH5wRA?e=KEk1zE'
+  # unzip downloaded folder to 2_observations/in
+  # such that the structure becomes
+  # 2_observations/in/reservoir_data
+      # WY2009 data/
+      # ...
+      # 01416900.WY2008.xls
+      # ...
+  2_observations/out/observed_daily_diversions.csv.ind:
+    command: read_munge_dailies(out_ind = target_name)
```
Comment (aappling-usgs) [inline 2_observations.yml:117]: Could be nice to "diversion" in the name of these functions (`read_munge_dailies`, `read_munge_monthlies`, `interpolate_to_daily`) so it's clear what they do when searching through functions

File: 2_observations.yml (hunk 1 above)
Comment (limnoliver) [reply 2_observations.yml:132]: Probably not -- but put here for reference if a user is wondering about units. Though, I made those explicit in the column names.

File: 2_observations/src/munge_diversion_data.R [hunk 3]
```diff
@@ -0,0 +1,239 @@
...
+    mutate(diversion_cms = diversion_cfs*cfs_to_cms) %>%
+    select(-storage_change_cfs, -diversion_cfs, -diversion_mgd)
+
+  # create a dataframe with all dates in timeseries for each reservoir
+  all_dates <- data.frame(date = seq(from = as.Date('1979-10-01'),
+                                          to = as.Date('2020-12-31'), by = 1)) %>%
+    mutate(year = lubridate::year(date),
+           month = lubridate::month(date, label = TRUE, abbr = FALSE))
+
+  all_dates <- bind_rows(mutate(all_dates, reservoir = 'Pepacton'), mutate(all_dates, reservoir = 'Cannonsville'))
```
Comment (aappling-usgs) [inline 2_observations/src/munge_diversion_data.R:147]: I recently learned about a cool feature of `bind_rows`, the `.id` argument, that could simplify this syntax if desired:
```r
> bind_rows(Pepacton=tibble(x=1:3), Cannonsville=tibble(x=4:6), .id='reservoir')
# A tibble: 6 x 2
  reservoir        x
  <chr>        <int>
1 Pepacton         1
2 Pepacton         2
3 Pepacton         3
4 Cannonsville     4
5 Cannonsville     5
6 Cannonsville     6
```

File: 2_observations/src/munge_diversion_data.R [hunk 4]
```diff
@@ -0,0 +1,239 @@
...
+  wy_path <- file.path(root_dir, sub_dirs)
+  wy_files <- list.files(wy_path)
+  wy_files <- wy_files[!grepl('csv', wy_files, ignore.case = TRUE)]
+
+  if(!length(wy_files) == 12) {
+    warning(paste(sub_dirs, ' does not contain 12 monthly files'))
+  }
+
+  dat2 <- wy_files %>%
+    map_df(~read_and_munge(location = wy_path, in_file = .x)) %>%
```
Comment (aappling-usgs) [inline 2_observations/src/munge_diversion_data.R:231]: Nice to see all this `purrr`ing - I've only recently come to really appreciate that package, but it's great

**Review by aappling-usgs — APPROVED** [review]
Looks great! Added some musings and cosmetic suggestions, but I'm happy for you to merge as-is if you need to move onto other tasks.

File: 2_observations/src/munge_diversion_data.R (hunk 3 above)
Comment (limnoliver) [reply 2_observations/src/munge_diversion_data.R:147]: Oh cool! I just learned the `.id` in `map_df` and used in this PR, very cool you can use it in `bind_rows` (and am betting `map_df` is just passing `.id` to `bind_rows`).

File: 2_observations/src/munge_diversion_data.R (hunk 4 above)
Comment (limnoliver) [reply 2_observations/src/munge_diversion_data.R:231]: This is the first time I've used it from scratch -- once I was writing a loop within a loop within a loop....it felt like the right time to make the jump. Still not totally intuitive to me, but love the resulting simplicity of the code!

**Pull request merged**